c++/src/capnp/capability.h

```diff
@@ -552,6 +560,39 @@ class Capability::Server {
   friend class LocalClient;
 };
 
+template <typename T>
+class RevocableServer {
+  // Allows you to create a capability client pointing to a capability server without taking
+  // ownership of the server. When `RevocableServer` is destroyed, all clients created through it
+  // will become broken. All outstanding RPCs via those clients will be canceled and all future
+  // RPCs will immediately throw. Hence, once the `RevocableServer` is destroyed, it is safe
+  // to destroy the server object it referenced.
+  //
+  // This is particularly useful when you want to create a capability server that points to an
+  // object that you do not own, and thus cannot keep alive beyond some defined lifetime. Since
+  // you cannot force the client to respect lifetime rules, you should use a RevocableServer to
+  // revoke access before the lifetime ends.
+  //
+  // The RevocableServer object can be moved (as long as the server outlives it).
+
+public:
+  RevocableServer(typename T::Server& server);
+  RevocableServer(RevocableServer&&) = default;
+  RevocableServer& operator=(RevocableServer&&) = default;
+  ~RevocableServer() noexcept(false);
+  KJ_DISALLOW_COPY(RevocableServer);
+
+  typename T::Client getClient();
+
+  void revoke();
```

Comment:
I'm very excited for this work! That said, can I convince you to remove `revoke()` without a message? Default generated kj exceptions have proven difficult to diagnose, especially with remote stacks.

Reply:
I'm expecting that the common case will be to revoke implicitly when the RevocableServer goes out-of-scope. Unfortunately there isn't really an opportunity to force the caller to provide a message there.

Reply:
I guess I can't argue after I made that change to kj::defer, huh? :sweat_smile: Perhaps that implies that there should be a default constructor as nullable type? Otherwise I'd expect people to just wrap it in a maybe. (And the movability does imply nullability)

Reply:
Sorry, I don't follow. How does that help improve the error message?

Reply:
It does not. Though perhaps we could provide a default error to the constructor? I was just commenting that the type was in a weird spot where it's a movable nullable (if we expect implicit behavior as common) but it's missing semantics.